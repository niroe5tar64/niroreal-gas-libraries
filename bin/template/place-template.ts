import fs from "fs";

const replaceText = (text: string, dict: Record<string, string>) => {
  const escapeRegExp = (str: string) => {
    return str.replace(/[.*+?^=!:${}()|[\]\/\\]/g, "\\$&");
  };
  const openParen = escapeRegExp("${");
  const closeParen = escapeRegExp("}");
  const keyRegexText = `${openParen}(${Object.keys(dict)
    .map(escapeRegExp)
    .join("|")})${closeParen}`;
  const keysRegex = new RegExp(keyRegexText, "g");
  return text.replace(keysRegex, (whole, key) => dict[key] ?? whole);
};

const placeTemplate = (templatePath: string, targetPath: string, dict: Record<string, string>) => {
  const template = fs.readFileSync(templatePath, "utf-8");
  const replaced = replaceText(template, dict);
  fs.writeFileSync(targetPath, replaced);
};

export { placeTemplate };
