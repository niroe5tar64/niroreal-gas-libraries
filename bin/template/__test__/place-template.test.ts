import { placeTemplate } from "../place-template";
import { describe, expect, test } from "bun:test";
import fs from "fs";
import path from "path";

describe("placeTemplate", () => {
  test("replaces the template with the given dictionary", () => {
    const __dirname = new URL(".", import.meta.url).pathname;
    const templatePath = path.resolve(__dirname, "sample-template.json");
    const targetPath = path.resolve(__dirname, "sample-target.json");
    const dict = {
      version: "0.0.1",
      name: "sample-file",
      date: new Date(2021, 0, 1).getTime().toString(),
    };
    placeTemplate(templatePath, targetPath, dict);

    const savedTarge = fs.readFileSync(targetPath, "utf-8");

    expect(savedTarge).toContain('"0.0.1"');
    expect(savedTarge).toContain('"sample-file"');
    expect(savedTarge).toContain("1609459200000");
    expect(savedTarge).toMatchSnapshot();
  });
});
