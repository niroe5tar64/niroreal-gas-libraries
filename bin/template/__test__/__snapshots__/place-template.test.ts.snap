// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`placeTemplate replaces the template with the given dictionary 1`] = `
"{
  "version": "0.0.1",
  "name": "sample-file",
  "date": 1609459200000
}"
`;
